design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,777 @@
+# End-to-end Testing in Crossplane
+
+* Owner: Predrag Knežević (@pedjak), Lovro Sviben (@lsviben)
+* Reviewers: @negz, TBD
+* Status: Draft
+
+## Background
+
+As a step in maturing Crossplane, we should be thinking about setting up
+processes and methodology that would enable automated end-to-end testing,
+covering all user-facing functionalities.
+
+Currently, as reported in [this issue], there seems to be a lack of e2e tests
+for Crossplane, impacting directly confidence in changes submitted in PRs and
+the quality of published releases.
+
+The following repos contains some sort of e2e tests:
+
+* [crossplane/test], last test change was 9 month ago
+* Crossplane repo itself [crossplane/crossplane e2e], last update was 10 months
+  ago
+* [crossplane/conformance], last change in June, 2021
+
+Apart for the stale development, the existing tests:
+
+* Cover only tiny fraction of all user-facing features
+* Do not cover any error cases
+* Are reported to be flaky sometimes
+
+The tests are written in pure Go, requesting to keep the number of library
+dependencies at minimum.
+
+Contrary to unit tests, end-to-end testing requires very often complex test 
+fixtures (e.g. installing Crossplane, providers, reaching some state in the
+cluster) before we can focus on validating a particular system behavior.
+Defining the fixture only with the help of Go standard library can be
+time-consuming, requiring to write a significant amount of boiler-plate code,
+hence reducing test readability and maintenance, and eventually not very
+motivating for creating new tests. In order to improve the readability, one
+would start looking for patterns and either eventually end up with a custom
+framework or adopt an existing one.
+
+**_NOTE_**: The document uses end-to-end testing term interchangeably
+with [system testing].
+
+## Goals
+
+Going forward we would like to explore and propose using an e2e testing
+framework, preferably one that would be already familiar to the community.
+The chosen framework should poses the following properties:
+
+* Support for declarative tests.
+
+  Similar to kubernetes object management, each test action should specify what
+  is the system state we would like to reach/check against. Such approach allows 
+  us to test effectively business rules exposed to users
+
+* Excellent readability for everybody (developers, users, Crossplane community
+  in general)
+
+  This would enable all stakeholders to have better/clear understanding about
+  the system behaviors.
```

Comment:
I suspect we're over-indexing on this goal. I might go so far as to question whether it - _excellent_ readability for non-developers - should be a goal at all. It feels to me like we're optimizing for the minority case at the expense of the majority case.

It certainly seems like a nice property, but at what cost? Is readability for non-developers so useful that we think contributors should need to learn not only a framework but a new DSL (i.e. Gherkin) just for E2E testing? This is essentially a decision to introduce a new technology into the project - not something we should take lightly.

I would expect that the _vast_ majority of the time the folks writing an E2E test will be the same folks designing, implementing, and unit testing the features that they'd like to E2E test. This isn't a project where we have a bunch of less-technical business stakeholders that need to validate that the software engineers built the thing to spec.

It's not _inconceivable_ to me that a community member who is not comfortable reading or writing Go would want to add an E2E test to demonstrate that something was broken, or want to take a look at an E2E test (instead of documentation) to try to understand how something works. I'm quite skeptical that it will happen often enough to warrant us introducing a new DSL though. I would expect the vast majority of folks reading and writing these E2E tests will be the same folks reading and writing the rest of the codebase - Go developers.

Reply:
Excellent readability is important for everybody. Tests code similar to production code is more often read than written. In order to enable developers to move fast and be productive, the readability plays a key role.

> It certainly seems like a nice property, but at what cost? 

If we want to measure costs, then we we need to come up with a set of measurable criteria, otherwise we can easily slip into subjective debate. Some possible criteria could be:

* Time to learn new concepts required for writing an end-to-end test
* Time to write an end-to-end test
* Time to modify single end-to-end test
* Time to modify a group of end-to-end tests

The proposed approach opens up possibility to produce useful impact beyond Crossplane validation. As mentioned in [Benefits Section](https://github.com/crossplane/crossplane/pull/4101/files#diff-e18ff1e97964f135a7e28848ac1355e73c999d953bd7563e0d39dadf8a07aebfR412), we could obtain testable documentation, where all snippets/example are tested as a part PR merge checks (cc @plumbis) 

> contributors should need to learn not only a framework but a new DSL (i.e. Gherkin) just for E2E testing? This is essentially a decision to introduce a new technology into the project - not something we should take lightly.

All evaluated frameworks are accompanied with a sort of DSL, in some cases it is delivered as a set of Go APIs/functions/conventions.

You can see Gherkin language that we control, as a sort of configuration file, but written in format that supports pure English sentences. IMHO, that is more readable than yaml, or json. As a part of this PR we have converted the exiting e2e test to Cucumber/godog approach. For that we needed to define just a handful of step/sentences, yet this is already enough for majority of tests that need to be written. 

A Crossplane contributor would need to learn following concepts:
* Test step name is just a much more readable name of Go function that gets invoked eventually
* Mapping between test step name and Go function is declared during test setup, in Go code
* Scenario description maps to test name, but again we can get more clarity due to free form text format
* Feature maps to test suite

Let's see how this works on example, [this test](https://github.com/crossplane/crossplane/pull/4101/files#diff-35a0aa68b3658cb9c9b0c3b7c17ba49d64b33709d8c4a0b88971157157ae1814R85-R193) is godog version of [existing `TestCompositeResourceGetsReady`](https://github.com/crossplane/crossplane/blob/master/test/e2e/apiextensions/composition_test.go#L48-L364) test:

* Pure Go tests has 3 times more lines of code
* Crafting/Applying CompositeResourceDefinition in Go looks like:

```go
&extv1.CompositeResourceDefinition{
	ObjectMeta: metav1.ObjectMeta{Name: "clusternopresources.nop.example.org"},
	Spec: extv1.CompositeResourceDefinitionSpec{
		Group: "nop.example.org",
		Names: kextv1.CustomResourceDefinitionNames{
			Kind:     "ClusterNopResource",
			ListKind: "ClusterNopResourceList",
			Plural:   "clusternopresources",
			Singular: "clusternopresource",
		},
		ClaimNames: &kextv1.CustomResourceDefinitionNames{
			Kind:     "NopResource",
			ListKind: "NopResourceList",
			Plural:   "nopresources",
			Singular: "nopresource",
		},
		ConnectionSecretKeys: []string{"test"},
		Versions: []extv1.CompositeResourceDefinitionVersion{{
			Name:          "v1alpha1",
			Served:        true,
			Referenceable: true,
			Schema: &extv1.CompositeResourceValidation{
				OpenAPIV3Schema: runtime.RawExtension{Raw: []byte(`{
					"type": "object",
					"properties": {
						"spec": {
							"type": "object",
							"properties": {
								"coolField": {
									"type": "string"
								}
							},
							"required": ["coolField"]
						}
					}
				}`)},
			},
		}},
	},
}
```

whereas `CompositeResourceDefinition is present` steps allows passing pure YAML content as the parameter:

```gherkin
CompositeResourceDefinition is present
      """
        apiVersion: apiextensions.crossplane.io/v1
        kind: CompositeResourceDefinition
        metadata:
          name: clusternopresources.nop.example.org
        spec:
          group: nop.example.org
          names:
            kind: ClusterNopResource
            listKind: ClusterNopResourceList
            plural: clusternopresources
            singular: clusternopresource
          claimNames:
            kind: NopResource
            listKind: NopResourceList
            plural: nopresources
            singular: nopresource
          connectionSecretKeys:
            - test
          versions:
            - name: v1alpha1
              served: true
              referenceable: true
              schema:
                openAPIV3Schema:
                  type: object
                  properties:
                    spec:
                      type: object
                      properties:
                        coolField:
                          type: string
                      required:
                        - coolField
```

IMHO, the later reads a way better.

* Similar to the above, step `claim becomes synchronized and ready` allows a reader to immediately understand the expectations, compared to the equivalent Go code:

```go
if err := wait.PollImmediate(10*time.Second, 90*time.Second, func() (done bool, err error) {
	d, err := dc.Resource(nopRes).Namespace("default").Get(context.TODO(), "nop-example", metav1.GetOptions{})
	if err != nil {
		t.Fatalf("Get nop-res %q: %v", nopresource.GetName(), err)
	}

	nopObj := composed.Unstructured{Unstructured: *d}
	isReady := nopObj.GetCondition(xpv1.TypeReady)

	if isReady.Status != corev1.ConditionTrue {
		t.Logf("nop-example %q is not yet Ready", nopresource.GetName())
		return false, nil
	}

	return true, nil
}); err != nil {
	t.Errorf("nop-example %q never became Ready: %v", nopresource.GetName(), err)
}
```

We could argue that the above code can be wrapped in a function and used like that, but IMHO, reading

`claim becomes synchronized and ready`

is much clearer then

`claimBecomesSynchronizedAndReady(context, claimName, namespace)`

A very important thing to state is that steps grammar is not dictated by Gherkin/godog - we are in full control of it, and are gonna design it according to our needs.

Reply:
First of all, I want to say that I'm impressed with the overall quality of the document and the e2e testing endeavor.

I agree with @pedjak that dealing with Golang structs instead of Yaml in the testing context is ugly. So let's please not do that, no matter the framework we pick in the end :D However, this comparison is orthogonal to the tooling, as we can always reference arbitrary Yaml files and include them in the tests, no matter the framework.

I have to also agree with @negz - the readability is a very subjective thing, so I can subjectively share that human-readable Cucumber/Gherkin is usually less **understandable** to me than structured Golang code or Rspec-like https://github.com/onsi/ginkgo code.

In my previous life, I was active in a Ruby community where Cucumber and Rspec were born, and the BDD movement was rising. Eventually, the same patterns were adopted by a wider community of languages and stacks. 

So I was doing Cucumber before it was cool(tm) in several projects and companies and can share some real-life stories. The original goal of Cucumber was/is to bring technical and non-technical stakeholders onto the same page. It is a very noble goal, but it happened in precisely 0 projects and companies where I witnessed(and in early cases was driving) the Cucumber adoption.

The end-state is usually the same: developers deal with the Gherkin abstraction they never need. Non-technical people are doing non-technical stuff and never contribute to the test suite. It makes the whole dance around human readability and step definitions an expensive redundant layer.

Back in the day, I preferred to stick with https://rspec.info/, which provided a middle ground: some form of BDD metastructure without much hiding the actual implementation. 

In the Golang world, there is https://github.com/onsi/ginkgo, a symmetric tool to Rspec. It was popular in the core Kubernetes testing effort.

As a concerning point, it is [mentioned](https://github.com/kubernetes-sigs/e2e-framework/blob/c3b933fe4c7313a31cbc2ea3cad2d6a2a3c8962f/docs/design/README.md?plain=1#L12) in e2e-framework documentation that moving away from Ginkgo is one of the primary motivation of the project creation.

Overall, to summarize my point:

* While Cucumber does look fancy and nice, Gherkin will be an almost useless and expensive layer long-term.
* The test creation process will contain serious overhead. Developers will have to think about the 'human readable' Gherkin representation string, then the Step definition regexp for it, and only then about actual test implementation. Again, too many layers. 
* Instead, I would love to have an original RSpec-like test layout where I can structure my test with BDD annotations and write implementation at the same place without many *mandatory* abstraction layers. Meanwhile, in this case, nothing stops me from abstracting things when it makes sense from a developer's standpoint.

This makes https://github.com/onsi/ginkgo another potential candidate to evaluate. It looks like people are successfully [using it ](https://itnext.io/testing-kubernetes-operators-with-ginkgo-gomega-and-the-operator-runtime-6ad4c2492379)to test standalone k8s operators.

The lightweight testing abstractions like [terratest](https://terratest.gruntwork.io/) ( evaluated in this doc) also [worked nicely](https://github.com/k8gb-io/k8gb/tree/master/terratest) for me in another project. 

I apologize if my points came across as strong. The overall doc and e2e testing goals are just excellent stuff!

Reply:
> The test creation process will contain serious overhead. Developers will have to think about the 'human readable' Gherkin representation string, then the Step definition regexp for it, and only then about actual test implementation. Again, too many layers.

Godog has a possibility to generate the up-to-date list of available steps with documentation as well, so developers can always see what are the steps they can use. Also, at beginning, while adding missing e2e tests, i.e. matching them with the available functionalities, there will be a high chance that new step needs to be introduced. However, after the test base stabilize, the experience shows that adding new step becomes very rare, and most of the time adding a new test requires just editing some of feature files.

Reply:
@negz 

>  Read and write Gherkin. (Yes it's English-language but it does appear to have some meaningful syntax).

The syntax is what we control and can make it as meaningful as we need.
> Read and write write Go code that parses Gherkin (e.g. ctx.Step('^claim becomes synchronized and ready$\', claimBecomesSynchronizedAndReady))

Nope, this is what Godog does for us - we only need to tell it what Go function to invoke when certain step name is matched. This is done once when a step gets introduced.

>    Read and write the Go code that seems to actually run the test, i.e. claimBecomesSynchronizedAndReady.

Not really needed, especially after the test code base stabilizes. At beginning, while adding missing e2e tests, i.e. matching them with the available functionalities, there will be a high chance that new step needs to be introduced. However, after the test base stabilize, the experience shows that adding new step becomes very rare, and most of the time adding a new test requires just editing some of feature files

Reply:
> Ginkgo is explicitly mentioned in [the contribution guide](https://github.com/crossplane/crossplane/blob/9ab0598dd964d796574dc5a1a712f4ee6687087f/contributing/README.md?plain=1#L50)

IIUC ginkgo is mentioned there in the unit test context so it does not invalidate using it in e2e scenario

Reply:
Disclaimer: I am one of the SIG Testing tech leads, primarily because of my work on the in-tree E2E framework.

> In the Golang world, there is https://github.com/onsi/ginkgo, a symmetric tool to Rspec. It was popular in the core Kubernetes testing effort.

I'd like to point out that we recently moved to Ginkgo v2, which addressed a lot of the deficiencies that were identified earlier by people using it for Kubernetes. In particular, we now have:
- Excellent support for defining timeouts and interrupting a running test suite, including running cleanup code after the tests have stopped.
- Intermediate progress reporting with details about what a test is waiting, if requested. This is useful for running tests interactively where waiting for a 5 minute timeout to get a failure message is extremely cumbersome.

See https://www.kubernetes.dev/blog/2023/04/12/e2e-testing-best-practices-reloaded/ for a summary of the recent changes.

Another enhancements that is partly implemented, partly coming include enabling linters to ensure that best practices are being followed.

> As a concerning point, it is [mentioned](https://github.com/kubernetes-sigs/e2e-framework/blob/c3b933fe4c7313a31cbc2ea3cad2d6a2a3c8962f/docs/design/README.md?plain=1#L12) in e2e-framework documentation that moving away from Ginkgo is one of the primary motivation of the project creation.

The direct consequences is that the e2e-framework now doesn't have any of the improvements above. They also don't support per-test log output from the underlying libraries (like client-go). This is annoying when running tests with `go test` (no `-v`) and the output for a failed test then doesn't include that klog output. It gets worse when running tests in parallel because the klog output from different tests gets mixed up.

The in-tree E2E framework is currently not recommended for usage outside of Kubernetes because depending on k/k is discouraged (too many indirect dependencies, not meant to be reused like that). It would be possible to change that by moving the E2E framework into a staging repo. However, this only makes sense for SIG Testing when it helps spread the work of maintaining that framework among more people. It doesn't make sense if we just get more issues filed against Kubernetes :sweat_smile: